force-app/main/default/classes/DMLOperations.cls

```diff
@@ -145,11 +163,47 @@ public with sharing class DMLOperations {
      * @param accountName The name of the Account.
      * @param oppNames The list of Opportunity names.
      */
+  
     public static void upsertOpportunities(String accountName, List<String> oppNames) {
+        Account acct;
+        
+        List<Account> accounts = [SELECT Id FROM Account WHERE Name = :accountName];
+        if (!accounts.isEmpty()) {
+            acct = accounts[0]; 
+        } else {
+            // If account isn't found, create and insert it
+            acct = new Account(Name = accountName);
+            insert acct;
+        }
+        
+
+        // Find existing opportunities for the account and add to Map: OppName => Opp
+        Map<String, Opportunity> existingOpps = new Map<String, Opportunity>();
+        List<Opportunity> exsistingOppsList = [SELECT Id, Name, StageName, CloseDate FROM Opportunity WHERE AccountId = :acct.Id AND Name IN :oppNames];
+        for (Opportunity opp : exsistingOppsList) {
```

Comment:
Try checking to see if the list has opps before starting the loop. You can stop the code from executing with a guard clause 
https://www.codementor.io/@clintwinter/use-guard-clauses-for-cleaner-code-1rrsczgwxp